Pointer dereference error when using connection string

### Expected Behavior
A connection to be established with Eventhub using environment variables: 
  EVENTHUB_CONNECTION_STRING, 
  EVENTHUB_NAMESPACE
  EVENTHUB_NAME

### Actual Behavior
The program fails when trying to find an aadProvider and never reaches the code to find an sasProvider. Stack trace below:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x121eafa]

goroutine 1 [running]:
eventhub/vendor/github.com/Azure/go-autorest/autorest/adal.retry(0x13022c0, 0xc420088f60, 0xc420124100, 0x0, 0x0, 0x12da6ec)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/go-autorest/autorest/adal/token.go:718 +0x18a
eventhub/vendor/github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).refreshInternal(0xc420122000, 0x1304e40, 0xc420016088, 0x12d3089, 0x1c, 0x0, 0x0)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/go-autorest/autorest/adal/token.go:653 +0x27f
eventhub/vendor/github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).RefreshWithContext(0xc420122000, 0x1304e40, 0xc420016088, 0x0, 0x0)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/go-autorest/autorest/adal/token.go:581 +0xa5
eventhub/vendor/github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).Refresh(0xc420122000, 0x35, 0x12d3089)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/go-autorest/autorest/adal/token.go:573 +0x43
eventhub/vendor/github.com/Azure/azure-amqp-common-go/aad.(*TokenProviderConfiguration).NewServicePrincipalToken(0xc42011c000, 0x0, 0x0, 0xc420014140)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/azure-amqp-common-go/aad/jwt.go:194 +0x5fd
eventhub/vendor/github.com/Azure/azure-amqp-common-go/aad.NewJWTProvider(0xc420057e80, 0x1, 0x1, 0xc420062040, 0xc420057ec0, 0xc420057e40)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/azure-amqp-common-go/aad/jwt.go:137 +0x126
eventhub/vendor/github.com/Azure/azure-event-hubs-go.NewHubWithNamespaceNameAndEnvironment(0xc42001c103, 0x18, 0xc42001414e, 0x4, 0x0, 0x0, 0x0, 0xc420010e00, 0xc420057f18, 0x105b3ad)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/azure-event-hubs-go/hub.go:139 +0x6f
eventhub/vendor/github.com/Azure/azure-event-hubs-go.NewHubFromEnvironment(0x0, 0x0, 0x0, 0x12329d7, 0xc420057f78, 0xc42007e058)
	/Users/conordarcy/Documents/go/src/eventhub/vendor/github.com/Azure/azure-event-hubs-go/hub.go:207 +0x156
main.main()
	/Users/conordarcy/Documents/go/src/eventhub/main.go:14 +0x32
exit status 2

### Environment
- OS: MacOS
- Go version: go version go1.10.2 darwin/amd64
- Version of Library: 0.3.0

Variables that are set:
EVENTHUB_NAMESPACE
EVENTHUB_NAME
EVENTHUB_CONNECTION_STRING

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -34,8 +34,8 @@
     "services/storage/mgmt/2017-10-01/storage",
     "version"
   ]
-  revision = "fad8443a79b0e755c18c3bec29a8d2bedab0b421"
-  version = "v15.3.0"
+  revision = "4650843026a7fdec254a8d9cf893693a254edd0b"
+  version = "v16.2.1"
 
 [[projects]]
   name = "github.com/Azure/azure-storage-blob-go"
@@ -191,6 +191,6 @@
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
-  inputs-digest = "24f9aedfb2f0a6e602f946d39cff1cb8900a337bedbc7ce18a3ca5b9334d2876"
+  inputs-digest = "9e8d39c2948233a4f9dc40bdfcbee21ae301de912713e0f0c8c6644668aece5b"
   solver-name = "gps-cdcl"
   solver-version = 1

--- a/Gopkg.toml
+++ b/Gopkg.toml
@@ -8,7 +8,7 @@
 
 [[constraint]]
     name = "github.com/Azure/azure-sdk-for-go"
-    version = "15"
+    version = "16"
 
 [[constraint]]
     name = "github.com/Azure/azure-amqp-common-go"

--- a/changelog.md
+++ b/changelog.md
@@ -1,5 +1,8 @@
 # Change Log
 
+## `v0.3.1`
+- simplify environmental construction by prefering SAS
+
 ## `v0.3.0`
 - pin version of amqp
 

--- a/hub.go
+++ b/hub.go
@@ -115,7 +115,6 @@ func NewHub(namespace, name string, tokenProvider auth.TokenProvider, opts ...Hu
 // There are two sets of environment variables which can produce a SAS TokenProvider
 //
 // 1) Expected Environment Variables:
-//   - "EVENTHUB_NAMESPACE" the namespace of the Event Hub instance
 //   - "EVENTHUB_KEY_NAME" the name of the Event Hub key
 //   - "EVENTHUB_KEY_VALUE" the secret for the Event Hub key named in "EVENTHUB_KEY_NAME"
 //
@@ -130,32 +129,24 @@ func NewHub(namespace, name string, tokenProvider auth.TokenProvider, opts ...Hu
 // 2. Client Certificate: attempt to authenticate with a Service Principal via "AZURE_TENANT_ID", "AZURE_CLIENT_ID",
 //    "AZURE_CERTIFICATE_PATH" and "AZURE_CERTIFICATE_PASSWORD"
 //
-// 3. Managed Service Identity (MSI): attempt to authenticate via MSI
+// 3. Managed Service Identity (MSI): attempt to authenticate via MSI on the default local MSI internally addressable IP
+//    and port. See: adal.GetMSIVMEndpoint()
 //
 //
 // The Azure Environment used can be specified using the name of the Azure Environment set in "AZURE_ENVIRONMENT" var.
 func NewHubWithNamespaceNameAndEnvironment(namespace, name string, opts ...HubOption) (*Hub, error) {
 	var provider auth.TokenProvider
-	aadProvider, aadErr := aad.NewJWTProvider(aad.JWTProviderWithEnvironmentVars())
-	sasProvider, sasErr := sas.NewTokenProvider(sas.TokenProviderWithEnvironmentVars())
-
-	if aadErr != nil && sasErr != nil {
-		// both failed
-		return nil, errors.Errorf("neither Azure Active Directory nor SAS token provider could be built - AAD error: %v, SAS error: %v", aadErr, sasErr)
+	provider, sasErr := sas.NewTokenProvider(sas.TokenProviderWithEnvironmentVars())
+	if sasErr == nil {
+		return NewHub(namespace, name, provider, opts...)
 	}
 
-	if aadProvider != nil {
-		provider = aadProvider
-	} else {
-		provider = sasProvider
+	provider, aadErr := aad.NewJWTProvider(aad.JWTProviderWithEnvironmentVars())
+	if aadErr == nil {
+		return NewHub(namespace, name, provider, opts...)
 	}
 
-	h, err := NewHub(namespace, name, provider, opts...)
-	if err != nil {
-		return nil, err
-	}
-
-	return h, nil
+	return nil, errors.Errorf("neither Azure Active Directory nor SAS token provider could be built - AAD error: %v, SAS error: %v", aadErr, sasErr)
 }
 
 // NewHubFromEnvironment creates a new Event Hub client for sending and receiving messages from environment variables

--- a/hub.go
+++ b/hub.go
@@ -44,7 +44,7 @@ const (
 	rootUserAgent   = "/golang-event-hubs"
 
 	// Version is the semantic version number
-	Version = "0.3.0"
+	Version = "0.3.1"
 )
 
 type (

--- a/hub_test.go
+++ b/hub_test.go
@@ -27,6 +27,7 @@ import (
 	"fmt"
 	"math/rand"
 	"os"
+	"strings"
 	"sync"
 	"testing"
 	"time"
@@ -50,8 +51,21 @@ var (
 	defaultTimeout = 20 * time.Second
 )
 
+const (
+	connStr  = "Endpoint=sb://namespace.servicebus.windows.net/;SharedAccessKeyName=keyName;SharedAccessKey=secret;EntityPath=hubName"
+)
+
 func TestEventHub(t *testing.T) {
 	suite.Run(t, new(eventHubSuite))
+}
+
+func (suite *eventHubSuite) TestNewHubWithNameAndEnvironment() {
+	revert := suite.captureEnv()
+	defer revert()
+	os.Clearenv()
+	suite.NoError(os.Setenv("EVENTHUB_CONNECTION_STRING", connStr))
+	_, err := NewHubWithNamespaceNameAndEnvironment("hello", "world")
+	suite.NoError(err)
 }
 
 func (suite *eventHubSuite) TestSasToken() {
@@ -96,6 +110,10 @@ func (suite *eventHubSuite) TestPartitioned() {
 
 	for name, testFunc := range tests {
 		setupTestTeardown := func(t *testing.T) {
+			//for _, e := range os.Environ() {
+			//	pair := strings.Split(e, "=")
+			//	fmt.Println(pair[0])
+			//}
 			hubName := suite.RandomName("goehtest", 10)
 			mgmtHub, err := suite.EnsureEventHub(context.Background(), hubName)
 			if err != nil {
@@ -460,9 +478,7 @@ func (suite *eventHubSuite) newClient(t *testing.T, hubName string, opts ...HubO
 func (suite *eventHubSuite) newClientWithProvider(t *testing.T, hubName string, provider auth.TokenProvider, opts ...HubOption) *Hub {
 	opts = append(opts, HubWithEnvironment(suite.Env))
 	client, err := NewHub(suite.Namespace, hubName, provider, opts...)
-	if err != nil {
-		t.Fatal(err)
-	}
+	suite.NoError(err)
 	return client
 }
 
@@ -484,4 +500,26 @@ func waitUntil(t *testing.T, wg *sync.WaitGroup, d time.Duration) {
 func fmtDuration(d time.Duration) string {
 	d = d.Round(time.Second) / time.Second
 	return fmt.Sprintf("%d seconds", d)
+}
+
+func restoreEnv(capture map[string]string) error {
+	os.Clearenv()
+	for key, value := range capture {
+		err := os.Setenv(key, value)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (suite *eventHubSuite) captureEnv() func() {
+	capture := make(map[string]string)
+	for _, pair := range os.Environ() {
+		keyValue := strings.Split(pair, "=")
+		capture[keyValue[0]] = strings.Join(keyValue[1:], "=")
+	}
+	return func() {
+		suite.NoError(restoreEnv(capture))
+	}
 }

--- a/internal/test/suite.go
+++ b/internal/test/suite.go
@@ -174,10 +174,7 @@ func (suite *BaseSuite) DeleteEventHub(ctx context.Context, name string) error {
 
 func (suite *BaseSuite) deleteAllTaggedEventHubs(ctx context.Context) {
 	client := suite.getEventHubMgmtClient()
-	res, err := client.ListByNamespace(ctx, ResourceGroupName, suite.Namespace)
-	if err != nil {
-		suite.FailNow(err.Error())
-	}
+	res, _ := client.ListByNamespace(ctx, ResourceGroupName, suite.Namespace)
 
 	for res.NotDone() {
 		for _, val := range res.Values() {
@@ -286,7 +283,7 @@ func (suite *BaseSuite) ensureNamespace() (*mgmt.EHNamespace, error) {
 }
 
 func (suite *BaseSuite) setupTracing() error {
-	if os.Getenv("CI") != "true" {
+	if os.Getenv("TRACING") == "true" {
 		// Sample configuration for testing. Use constant sampling to sample every trace
 		// and enable LogSpan to log every span via configured Logger.
 		cfg := config.Configuration{

